File: MatlabCode/Repo_CompIntroQP_MATLAB/ChapterQtech/Exercise_6_1_1_f.m
% This script determines the tunneling rate for a particle of unit mass 
% hitting a potential supported in a finite interval.
% It solves the time-independent Schrödinger equation in the region where 
% the potential is supported by formulating it as a first order coupled 
% ODE in \psi(x) and \phi(x), where \phi(x) = \psi'(x). It does so for a 
% set of several energy values.
%
% Inputs:
% Vpot  - The potential; this input is an inline function - one
% that features certain additional input parameters
% D     - The width of the interval in which the potential 
% is nonzero
% Emin  - The minimal energy of the particle, must be positive
% Emax  - The maximal energy
% dE    - The increment of the vector containing input energies
% 
% These input parameters are hard coded initially.

% Input parameters
D = 3;
Emin = 0.1;
Emax = 4;
dE = 0.05;

% The potential
V0 = 2;
a = 0.4;
Vpot = @(x) V0-a*x;

% Allocate/initiate vectors
EnergyVector = Emin:dE:Emax;
len = length(EnergyVector);
TransVector = zeros(1, len);

index = 1;
for Energy = EnergyVector
  TransVector(index) = TransProb(Energy, Vpot, D);
  index = index + 1;
end

% Plot result
figure(1)
subplot(1,2,1)
plot(EnergyVector, TransVector, 'k-', 'linewidth', 2)
set(gca, 'fontsize', 20)
xlabel('Energy')
ylabel('Transmission Probability')
grid on
subplot(1,2,2)
semilogy(EnergyVector, TransVector, 'k-', 'linewidth', 2)
set(gca, 'fontsize', 20)
xlabel('Energy')
ylabel('Transmission Probability')
grid on

% Function which calculates T for a given energy
function T = TransProb(E, Vpot, D)

  % Wave number
  k = sqrt(2*E);

  % Coefficient matrix for ODE
  % ODE
  RHS = @(x,y) [0 1; 2*Vpot(x)-k^2 0]*y;
  % "Initial value"
  yD = [1; 1i*k];
  % Solve ODE
  [x y] = ode45(RHS, [D 0], yD);

  % Read off psi(0) and psi'(0) = phi(0)
  Psi0 = y(end, 1);
  Phi0 = y(end, 2);
  % Transmission probability
  T = 4*k^2/abs(k*Psi0-1i*Phi0)^2;
end

File: PythonCode/Chapter6_QuantumTech/Exercise_6_1_1_f.py
"""
 This script determines the tunneling rate for a particle of unit mass 
 hitting a potential supported in a finite interval.
 It solves the time-independent Schrödinger equation in the region where 
 the potential is supported by formulating it as a first order coupled ODE in 
 \psi(x) and \phi(x), where \phi(x) = \psi'(x). It does so for a set of 
 several energy values.

 Inputs:
 Vpo   - The potential this input is a function variable- one
 that features certain additional input parameters
 D     - The width of the interval in which the potential 
 is nonzero
 Emin  - The minimal energy of the particle, must be positive
 Emax  - The maximal energy
 dE    - The increment of the vector containing input energies
 
 These input parameters are hard coded initially.
"""

# Libraries
import numpy as np
from matplotlib import pyplot as plt
from scipy import integrate

# Input parameters
D = 3
Emin = 0.1
Emax = 4
dE = 0.05

# The potential
V0 = 2
a = 0.4
def Vpot(x):
    return V0-a*x


# Function which calculates T for a given energy
def TransProb(E):
    # Wave number
    k = np.sqrt(2*E)

    # Set up the equation the right hand side of the ODE,
    # y'(t) = M y(t), where y is [psi, phi] and M is
    # the coefficient matrix
    def RHS(x, y):
        CoeffMat = np.matrix([[0, 1], [2*Vpot(x)-k**2, 0]])
        return np.matmul(CoeffMat, y.reshape(2,1))
      
    # Initial state (must be  complex)
    y0 = [complex(1,0), 1j*k]
    y0 = np.asarray(y0)

    # ODE solver
    # Numerical solution of the ODE
    sol = integrate.solve_ivp(RHS, (D, 0), y0, vectorized = True)

    # Get psi(0) and phi(0)
    psi0 = sol.y[0, -1]
    phi0 = sol.y[1, -1]

    # Numerical transmission probability
    T = 4*k**2/np.abs(k*psi0-1j*phi0)**2
    return T


# Allocate/initiate vectors
EnergyVector = np.arange(Emin, Emax+dE, dE)
L = len(EnergyVector)
TransVector = np.zeros(L)

index = 0
for Energy in EnergyVector:
  TransVector[index] = TransProb(Energy)
  index = index + 1

# Plot result
plt.figure(1)
plt.clf()
# With linear axes
plt.plot(EnergyVector, TransVector, '-', linewidth = 2, 
         color = 'black')
# With logarithmic y-axis
#plt.semilogy(EnergyVector, TransVector, '-', linewidth = 2, 
#         color = 'black')
plt.xlabel('Energy', fontsize = 12)
plt.ylabel('Transmission Probability', fontsize = 12)
plt.grid()
plt.show()